src/utils/proxyMap.ts

```diff
@@ -0,0 +1,114 @@
+import { proxy } from 'valtio'
+
+type InternalProxyMap<K, V> = Map<K, V> & {
+  data: {
+    [key: string]: V
+  }
+  toJSON: object
+  makeMap(): Map<K, V>
+}
+
+export const proxyMap = <K, V>(entries?: [K, V][] | null): Map<K, V> => {
+  // store user key and dataRef used to access the value provided by the user
+  const keyStore = new Map<K, number>()
```

Comment:
I think having something outside `map` will break the behavior with `snapshot`.
The magic how proxySet/Map work is we translate Set/Map to plain objects(arrays).